src/main/java/teammates/common/util/TzdbResourceZoneRulesProvider.java

```diff
@@ -0,0 +1,181 @@
+package teammates.common.util;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.StreamCorruptedException;
+import java.lang.reflect.Method;
+import java.time.zone.ZoneOffsetTransition;
+import java.time.zone.ZoneOffsetTransitionRule;
+import java.time.zone.ZoneRules;
+import java.time.zone.ZoneRulesException;
+import java.time.zone.ZoneRulesProvider;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Loads time-zone rules for 'TZDB'. Modified from {@link java.time.zone.TzdbZoneRulesProvider}.
+ * Reads the file 'tzdb.dat' from resources instead of the JRE library directory.
+ */
+public final class TzdbResourceZoneRulesProvider extends ZoneRulesProvider {
+
+    /** Type for ZoneRules. */
+    private static final byte ZRULES = 1;
+    /** Type for ZoneOffsetTransition. */
+    private static final byte ZOT = 2;
+    /** Type for ZoneOffsetTransition. */
+    private static final byte ZOTRULE = 3;
+
+    /**
+     * All the regions that are available.
+     */
+    private List<String> regionIds;
+    /**
+     * Version Id of this tzdb rules.
+     */
+    private String versionId;
+    /**
+     * Region to rules mapping.
+     */
+    private final Map<String, Object> regionToRules = new ConcurrentHashMap<>();
+
+    /**
+     * Creates an instance.
+     * Created by the {@code ServiceLoader}.
+     *
+     * @throws ZoneRulesException if unable to load
+     */
+    public TzdbResourceZoneRulesProvider() {
```

Comment:
Can this be `private`?

Reply:
Nope; `registerResourceZoneRules` will fail at runtime if it is.